### PR TITLE
Download to File Investigation

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Replaces `IOUtils` writing of the download stream to file to using `FluxUtils.writeToAsynchronousByteChannel` as a bug was found with the terminal API `IOUtils` was using when there are random network errors while downloading the stream.

At the time of this fix `NettyAsyncHttpResponse.writeBodyToAsync` had a race condition where `IOUtils.transferStreamResponseToAsynchronousByteChannel` would track the progress of writing the network response to the `AsynchronousByteChannel` and a network error would have the request retried from the offset where the error occurred. Sinc `AsynchronousByteChannel` writes on another thread and Reactor will eagerly propagate `onError` signals after `onNext` signals without the requirement for the subscriber to request another emission it could lead to a scenario where the `onNext` write began being handled by the channel write thread, `IOUtils.transferStreamResponseToAsynchronousByteChannel` saw the `onError` and captured the write state location, writing to file completed updating the write to channel position, and finally the request was retried with a stale position. An explicit example would be:

- Request made to download blob range 0-8192
- Network emits chunk 0-4096
- Writing of the chunk begins
- Network has an error and the request is retried using 0-8196 as the writing offset position was never updated
- Write of chunk completes updating channel position to 4092
- Response for retry request is returned with range 0-8192
- Writing of chunk begins at 4096
- Resulting channel has data 0-4096 and 0-8192 written resulting in range 0-4096 being written twice.

Switching to `FluxUtils.writeToAsynchronousByteChannel` mitigates this issue as `FluxUtils.writeToAsynchronousByteChannel` is handled much further down the call stack than `IOUtils.transferStreamResponseToAsynchronousByteChannel` where `onNext` channel writing and `onError` request retrying can't happen at the same time.

No other SDKs have been found using the affected code path that is no longer being called in the `azure-storage-blob` SDK.

See #35004 for the fix to `NettyAsyncHttpResponse.writeBodyToAsync`.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java [hunk 1]
```diff
@@ -93,10 +92,12 @@ private static Flux<ByteBuffer> createResponseFlux(StreamResponse sourceResponse
...
     public Mono<Void> writeValueToAsync(AsynchronousByteChannel channel, ProgressReporter progressReporter) {
         Objects.requireNonNull(channel, "'channel' must not be null");
-        if (sourceResponse != null) {
-            return IOUtils.transferStreamResponseToAsynchronousByteChannel(channel, sourceResponse, onErrorResume,
-                progressReporter, retryOptions.getMaxRetryRequests());
-        } else if (super.getValue() != null) {
+//        if (sourceResponse != null) {
+//            return IOUtils.transferStreamResponseToAsynchronousByteChannel(channel, sourceResponse, onErrorResume,
+//                progressReporter, retryOptions.getMaxRetryRequests());
+//        }
```
Comment (ibrahimrabab) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java:98]: Remove the commented code

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java [hunk 2]
```diff
@@ -51,8 +50,8 @@ public BlobDownloadAsyncResponse(HttpRequest request, int statusCode, HttpHeader
         BlobDownloadHeaders deserializedHeaders) {
         super(request, statusCode, headers, value, deserializedHeaders);
         this.sourceResponse = null;
-        this.onErrorResume = null;
-        this.retryOptions = null;
+        //this.onErrorResume = null;
+        //this.retryOptions = null;
```
Comment (ibrahimrabab) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java:54]: Remove the commented code

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java [hunk 3]
```diff
@@ -67,8 +66,8 @@ public BlobDownloadAsyncResponse(HttpRequest request, int statusCode, HttpHeader
         super(sourceResponse.getRequest(), sourceResponse.getStatusCode(), sourceResponse.getHeaders(),
             createResponseFlux(sourceResponse, onErrorResume, retryOptions), extractHeaders(sourceResponse));
         this.sourceResponse = Objects.requireNonNull(sourceResponse, "'sourceResponse' must not be null");
-        this.onErrorResume = Objects.requireNonNull(onErrorResume, "'onErrorResume' must not be null");
-        this.retryOptions = Objects.requireNonNull(retryOptions, "'retryOptions' must not be null");
+        //this.onErrorResume = Objects.requireNonNull(onErrorResume, "'onErrorResume' must not be null");
+        //this.retryOptions = Objects.requireNonNull(retryOptions, "'retryOptions' must not be null");
```
Comment (ibrahimrabab) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java:70]: Remove the commented code

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java:98]: I'll add comments as we'll want to keep this once the issue is resolved in the lower levels as it'll allow us to reintroduce performance improvements.

**Review by ibrahimrabab — APPROVED** [review]

**Pull request merged**